plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -734,12 +772,65 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+        if (connectorTablePartitioning.isEmpty()) {
+            return Optional.empty();
+        }
+        IcebergTablePartitioning tablePartitioning = connectorTablePartitioning.get();
+
+        // Check if the table can be partitioned on the requested columns
+        if (!new HashSet<>(tablePartitioning.partitioningColumns()).containsAll(partitioningColumns)) {
+            return Optional.empty();
+        }
+
+        Map<ColumnHandle, Integer> newPartitioningColumnIndex = IntStream.range(0, partitioningColumns.size()).boxed()
+                .collect(toImmutableMap(partitioningColumns::get, identity()));
+        ImmutableList.Builder<IcebergPartitionFunction> newPartitionFunctions = ImmutableList.builder();
+        ImmutableList.Builder<Integer> newPartitionStructFields = ImmutableList.builder();
+        for (int functionIndex = 0; functionIndex < tablePartitioning.partitioningHandle().partitionFunctions().size(); functionIndex++) {
+            IcebergPartitionFunction function = tablePartitioning.partitioningHandle().partitionFunctions().get(functionIndex);
+            int oldColumnIndex = function.dataPath().getFirst();
+            Integer newColumnIndex = newPartitioningColumnIndex.get(tablePartitioning.partitioningColumns().get(oldColumnIndex));
+            if (newColumnIndex != null) {
+                // Change the index of the top level column to the location in the new partitioning columns
+                newPartitionFunctions.add(function.withTopLevelColumnIndex(newColumnIndex));
+                // Some partition functions may be dropped so update the struct fields used in split partitioning must be updated
+                newPartitionStructFields.add(tablePartitioning.partitionStructFields().get(functionIndex));
+            }
+        }
+
+        IcebergPartitioningHandle newPartitioningHandle = new IcebergPartitioningHandle(false, newPartitionFunctions.build());
+        if (partitioningHandle.isPresent() && !partitioningHandle.get().equals(newPartitioningHandle)) {
+            // todo if bucketing is a power of two, we can adapt the bucketing
+            return Optional.empty();
+        }
+        if (newPartitioningHandle.partitionFunctions().stream().map(IcebergPartitionFunction::transform).noneMatch(BUCKET::equals)) {
+            // The table is only using value-based partitioning, and this can hurt performance if there is a filter
+            // on the partitioning columns. This is something we may be able to support with statistics in the future.
+            return Optional.empty();
+        }
+
+        return Optional.of(icebergTableHandle.withTablePartitioning(Optional.of(new IcebergTablePartitioning(
+                true,
+                newPartitioningHandle,
+                partitioningColumns.stream().map(IcebergColumnHandle.class::cast).collect(toImmutableList()),
+                newPartitionStructFields.build()))));
+    }
+
```

Comment:
I'm assuming lack of `getCommonPartitioningHandle` implementation here just means that bucket count mismatch cases are not handled yet ?

Reply:
Mismatch is not supported.  There is a comment in the applyPartitioning method